### ISSUE TITLE
Build out metrics support on OpenTelemetry.Instrumentation.SqlClient and OpenTelemetry.Instrumentation.EntityFrameworkCore

### ISSUE THREAD

**Issue by kharabasz** [open-telemetry/opentelemetry-dotnet-contrib#2106]
### Component

OpenTelemetry.Instrumentation.SqlClient

### Is your feature request related to a problem?

Currently metrics support for:
```
"Microsoft.Data.SqlClient.EventSource" // https://github.com/dotnet/SqlClient/blob/v5.1.5/src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlClientEventSource.cs"
"Microsoft.EntityFrameworkCore" // https://github.com/dotnet/efcore/blob/v8.0.3/src/EFCore/Infrastructure/EntityFrameworkEventSource.cs#L45
```
is available via the `OpenTelemetry.Instrumentation.EventCounters` package.

The implementation there is meant to be generic and has some gotchas, namely around how instruments names are constructed. Metrics from the same event source can end up with hard to parse instrument name, for example:
```
ec.Microsoft.EntityFrameworkCore.active-db-contexts
ec.Microsoft.Entity.total-execution-strategy-operation-failures
ec.Microsoft.E.execution-strategy-operation-failures-per-second-total
ec.Microsoft.EntityFramew.total-optimistic-concurrency-failures
```

While rewrites this can be managed with views, the filtering code becomes involved:
```
        var otel = builder.Services.AddOpenTelemetry();

        var eventCounterMetricsToBeRenamed = new Dictionary<string, string>
        {
            ["active-hard-connections"] = "dotnet_sqlclient_active_hard_connections",
            ["number-of-non-pooled-connections"] = "dotnet_sqlclient_number_of_non_pooled_connections",
            ["number-of-pooled-connections"] = "dotnet_sqlclient_number_of_pooled_connections",
            ["number-of-stasis-connections"] = "dotnet_sqlclient_number_of_stasis_connections",
            ["number-of-reclaimed-connections"] = "dotnet_sqlclient_number_of_reclaimed_connections",
            ["active-db-contexts"] = "dotnet_ef_active_db_contexts",
            ["queries-per-second"] = "dotnet_ef_queries_per_second",
            ["execution-strategy-operation-failures-per-second"] = "dotnet_ef_execution_strategy_operation_failures_per_second",
            ["optimistic-concurrency-failures-per-second"] = "dotnet_ef_optimistic_concurrency_failures_per_second",
        };

        otel.WithMetrics(meterProvider => meterProvider
            .AddEventCountersInstrumentation(options => // https://learn.microsoft.com/en-us/dotnet/core/diagnostics/event-counters
            {
                options.AddEventSources("Microsoft.Data.SqlClient.EventSource"); // https://github.com/dotnet/SqlClient/blob/v5.1.5/src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlClientEventSource.cs#L73
                options.AddEventSources("Microsoft.EntityFrameworkCore"); // https://github.com/dotnet/efcore/blob/v8.0.3/src/EFCore/Infrastructure/EntityFrameworkEventSource.cs#L45
            })
            .AddView(instrument =>
            {
                // EventCounters instrumentation adds a ton of noise by default. We need to filter down to the metrics we care about.
                if (instrument.Meter.Name == "OpenTelemetry.Instrumentation.EventCounters")
                {
                    // EventCounters instrument names follow the format "ec.Some.Event.Namespace.metric-name"
                    // We can't filter by namespace as it gets truncated and is not capture elsewhere on the instrument
                    var metricName = instrument.Name.Split(".").LastOrDefault();
                    if(!string.IsNullOrEmpty(metricName) && eventCounterMetricsToBeRenamed.TryGetValue(metricName, out var selectedName))
                    {
                        return new MetricStreamConfiguration { Name = selectedName };
                    }

                    // Drop anything else to control costs
                    return MetricStreamConfiguration.Drop;
                }

                return null;
            })
            .AddPrometheusExporter(o => o.DisableTotalNameSuffixForCounters = true) // fixes known issue: https://github.com/open-telemetry/opentelemetry-dotnet-contrib/issues/1617
        );
```

Managing the filter/rewrite list at the code level becomes hard in organizations with a large number of applications. In our current OpenTelemetry setup, we manage this filtering/rewrite logic globally on our ingestion gateway.

### What is the expected behavior?

Gathering SqlClient and EntityFrameworkCore stats seems like a common use case. It would be great if the already existing [OpenTelemetry.Instrumentation.SqlClient](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/tree/main/src/OpenTelemetry.Instrumentation.SqlClient) and [OpenTelemetry.Instrumentation.EntityFrameworkCore](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/tree/main/src/OpenTelemetry.Instrumentation.EntityFrameworkCore) packages supported metrics for their corresponding event sources in the same way as done by the [OpenTelemetry.Instrumentation.Runtime](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/tree/main/src/OpenTelemetry.Instrumentation.Runtime) package. 

This would allow the libraries to produce instruments with a standard prefix, which would be easier to filter or rewrite in telemetry ingestion pipelines.

### Which alternative solutions or features have you considered?

Views, as mentioned in the feature request.

**Comment by Kielek** [comment]
@kharabasz, please check #2026. If you have feedback/contribution, it is more than welcomed.

**Comment by cijothomas** [comment]
Can be closed as duplicate of https://github.com/open-telemetry/opentelemetry-dotnet-contrib/issues/1738
Piotr also shared the WIP PR for the sqlclient metrics.

**Comment by shmuntz** [comment]
Is there a way to collect additional EventCounter metrics from SQL Client, like "number-of-non-pooled-connections" or "number-of-pooled-connections"? I see that current OpenTelemetry.Instrumentation.SqlClient package supports only the "db.client.operation.duration" histogram, but what about many others? Is there a ticket open for this feature request? If someone knows anything about such plans for the future release, please post a response here.
